src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java

```diff
@@ -1977,37 +1977,42 @@ final void runWorker(WorkQueue w) {
                     int j, cap; WorkQueue q; ForkJoinTask<?>[] a;
                     if ((q = qs[j = i & (n - 1)]) != null &&
                         (a = q.array) != null && (cap = a.length) > 0) {
-                        for (int pb = -1, b; ; pb = b) {  // track progress
-                            ForkJoinTask<?> t; int nb; long k, nk;
-                            t = (ForkJoinTask<?>)U.getReferenceAcquire(
-                                a, k = slotOffset((cap - 1) & (b = q.base)));
-                            Object u = U.getReference(    // next slot
-                                a, nk = slotOffset((cap - 1) & (nb = b + 1)));
+                        for (int m = cap - 1, pb = -1, b; ; pb = b) {
+                            long k = slotOffset((b = q.base) & m);
+                            ForkJoinTask<?> t = (ForkJoinTask<?>)
+                                U.getReferenceAcquire(a, k);
                             if (q.base != b)              // inconsistent
                                 ;
                             else if (t == null) {
-                                if (u == null && q.top - b <= 0)
-                                    break;                // empty
+                                if (rescan)
+                                    break;                // will revisit
+                                if (a[b & m] == null &&
+                                    a[(b + 1) & m] == null &&
+                                    a[(b + 2) & m] == null)
+                                    break;                // probably empty
                                 if (pb == b) {
-                                    rescan = true;        // stalled; revisit
+                                    rescan = true;        // stalled; reorder scan
                                     break scan;
                                 }
                             }
                             else if (U.compareAndSetReference(a, k, t, null)) {
-                                q.base = nb;
+                                q.base = b + 1;
                                 w.nsteals = ++nsteals;
                                 w.source = j;             // volatile
-                                rescan = true;
-                                if (src != (src = j) &&
-                                    U.getReferenceVolatile(a, nk) != null)
+                                boolean propagated = taken;
+                                rescan = taken = true;
+                                if ((!propagated ||
+                                     ((j & 1) == 0) && t instanceof
+                                     ForkJoinTask.InterruptibleTask) &&
```

Comment:
It's probably premature, but given that all subclasses of InterruptibleTask are final, and that InterruptibleTask is abstract, it _might_ be cheaper to check `t.getClass().getSuperClass() == ForkJoinTask.InterruptibleTask.class`? 🤔